### PR TITLE
created thread-safe lazy initialization using Memoizer class

### LINKED ISSUE #1015: WindowsUsbDevice#getControllerDeviceIdList is not thread safe
I'm marking this issue for [first-timers-only](https://www.firsttimersonly.com/).  That means that I will only accept a PR for this one from someone who's never contributed to open source before. This one is easy (but don't make that statement make you feel bad if you have a hard time with it, there's more to contributing to open source than changing lines of code, especially if it's your first time). I'll hold your hand through this if you need me to. :-) 

As noted in [this report](https://scrutinizer-ci.com/g/oshi/oshi/issues/master/files/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java?selectedLabels%5B0%5D=37&orderField=path&order=asc&honorSelectedPaths=0&issueId=40218079#inspectioncomment-136359521), the lazy initialization of a list in `WindowsUSBDevice` is not thread safe.  If two threads call `getUsbDevices()` simultaneously they could both pass the `null` check and then both attempt to modify the same variable.

Create a thread-safe lazy initialization using OSHI's `Memoizer` class.

Here are the steps to get a PR merged here.
- Read the [CONTRIBUTING](https://github.com/oshi/oshi/blob/master/CONTRIBUTING.md) file for reference.  I'll walk you through anything you don't understand.
- Fork the project to your own GitHub account, make a local clone, and create a branch for your submission.
- In the [`WindowsUsbDevice`](https://github.com/oshi/oshi/blob/master/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java) class:
  - Modify the [`getControllerDeviceIdList()`](https://github.com/oshi/oshi/blob/31858912752ada0e0ceb43b43196ed1409601b96/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java#L309) method to take no arguments.  Inside the method, get the `wmiQueryHandler` by [copying from line 136](https://github.com/oshi/oshi/blob/31858912752ada0e0ceb43b43196ed1409601b96/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java#L136).  
  - Also remove the `if (controllerDeviceIds == null)` conditional 
  - Redefine that `controllerDeviceIds` variable with a differently named local variable in the method. 
  - Replace the static class variable with a memoized supplier: `private static Supplier<List<String>> controllerDeviceIds = Memoizer
            .memoize(WindowsUsbDevice::getControllerDeviceIdList);`
  - In the [`getUsbDevices()` method](https://github.com/oshi/oshi/blob/31858912752ada0e0ceb43b43196ed1409601b96/oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java#L137) instead of calling the old method, instead fetch from the memoizer: `List<String> controllerDeviceIdList = controllerDeviceIds.get();`
- Commit your changes, push to to your GitHub fork, and create a Pull Request
- Respond to the usually minor comments on code review
- Celebrate! :tada:

### PR DESCRIPTION
**Pull request by DanielMK01**
Thank you for this opportunity! I look forward to hearing from you soon!

Fixes #1015

### CODE REVIEW

File: oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java [hunk 1]
```diff
@@ -55,269 +57,256 @@
  */
 public class WindowsUsbDevice extends AbstractUsbDevice {
 
-    private static final Logger LOG = LoggerFactory.getLogger(WindowsUsbDevice.class);
+	private static final Logger LOG = LoggerFactory.getLogger(WindowsUsbDevice.class);
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: You may not have switched your formatter to use the OSHI conventions, as all your indents have turned into tabs.

File: oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java [hunk 2]
```diff
@@ -55,269 +57,256 @@
...
-    public WindowsUsbDevice(String name, String vendor, String vendorId, String productId, String serialNumber,
-            String uniqueDeviceId, UsbDevice[] connectedDevices) {
-        super(name, vendor, vendorId, productId, serialNumber, uniqueDeviceId, connectedDevices);
-    }
+	/**
+	 * <p>
+	 * Constructor for WindowsUsbDevice.
+	 * </p>
+	 *
+	 * @param name             a {@link java.lang.String} object.
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: This is another formatting convention that OSHI uses with the newline/indent for parameters.

File: oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java [hunk 3]
```diff
@@ -55,269 +57,256 @@
...
-                controllerDeviceIds.add(WmiUtil.getString(usbController, USBControllerProperty.PNPDEVICEID, i));
-            }
-        }
-        return controllerDeviceIds;
-    }
+	/**
+	 * Queries the USB Controller list, which doesn't change so we only need to
+	 * query it once
+	 *
+	 * @param wmiQueryHandler
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: Since you've deleted the method parameter, you should also delete the `@param` javadoc.

File: oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java (hunk 1 above)
Comment (DanielMK01) [reply oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: Thanks for reminding me! I'll try and fix it right now, and I'll let you know how it goes.

Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: All you need to do when you fix it is to push the new commit, and it'll automatically update this request.

File: oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java [hunk 4]
```diff
@@ -55,269 +57,256 @@
...
-        PNPDEVICEID;
-    }
+	enum USBControllerProperty {
+		PNPDEVICEID;
+	}
 
-    // List of controllers
-    private static List<String> controllerDeviceIds = null;
+	// List of controllers
+	private static Supplier<List<String>> controllerDeviceIds = Memoizer.memoize(WindowsUsbDevice::getControllerDeviceIdList);
```
Comment (dbwiddis) [inline oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: One more small request.  Elsewhere in OSHI I've done a static import of the `Memoizer.memoize` method to shorten this to just `memoize(...)`.  Would be nice to do that here for consistency.

Comment (DanielMK01) [reply oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: I'll go fix that right now. I just reformatted the code to the OSHI standard, but every time I try to commit the new code, it says "Changes not staged for commit:" Can you help me with this issue?

Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: You need to commit the changed files!  Command line is `git commit` but there should be a menu option in Eclipse under the "Team" menu.

Comment (DanielMK01) [reply oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: Also, once I remove the "Memoizer.", the code no longer compiles. The error says "The method memoize(WindowsUsbDevice::getControllerDeviceIdList) is undefined for the type WindowsUsbDevice" Will this be an issue?

Comment (dbwiddis) [reply oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: You have to add an `import static oshi.util.Memoizer.memoize;` statement.

Comment (DanielMK01) [reply oshi-core/src/main/java/oshi/hardware/platform/windows/WindowsUsbDevice.java]: Alright, I believe the changes have been committed and pushed. Can you verify this for me please?

**Comment by dbwiddis** [comment]
This looks great!  I'll let the tests finish and then merge!

**Comment by DanielMK01** [comment]
Wow that was intense! I learned a lot today thanks to you. I'll be signing off now, so have a good night!
